### PR TITLE
Update specs to OpenAPI 3.0.1

### PR DESCRIPTION
**Pull request by clarakosi**
Dependent on[ Hyperswitch PR](https://github.com/wikimedia/hyperswitch/pull/101)

I should’ve caught this earlier but the online swagger validator we were using does not support v3. 

Instead of waiting for swagger-cli to update its dependency (swagger-parser using  openapi-schema-validation v0.4.2 instead of v3.0.3) I’ve elected to use openapi-schema-validation v3.0.3 directly. I am open to other suggestions.

Summary of major changes in v3:
1. `swagger: 2.0` has been replaced with `openapi: 3.0.1`
    1. https://swagger.io/docs/specification/basic-structure/
2. `definitions` was renamed to `schemas` and `securityDefinitions`  to `securitySchemes` and they all were moved inside components
    1. https://swagger.io/docs/specification/components/
3. `servers` replaces the `basePath` keywords used in OpenAPI 2.0
    ~~1. **ISSUE/ Additional work:** Service checker expects a basePath. It needs to be updated to also look for servers[0].url~~
    2. https://github.com/wikimedia/operations-software-service-checker/blob/master/servicechecker/swagger.py#L109
4. `produces` replaced with `content` in `response`. Must include media type & schema. 
    1. https://swagger.io/docs/specification/media-types/
5. Examples in request parameters are back (originally removed when converting to v2.0)
6. `consumes` has been replaced with `requestBody`
    1. https://swagger.io/docs/specification/describing-request-body/
7. Data type: File has been replace with type: string, format: binary
    1. https://swagger.io/docs/specification/data-models/data-types/
8. Reordering of `type` in `parameters` to show up under `schema`

~~**Unclear/ Need help:**
I’m not fully sure how redirect works in the codebase. For example, with `html/{title}` && `html/{title}/{revision}` we have it accepting a boolean (in spec) but some of the tests are providing the value `no` to the query. I’ve gone through some of the code in sys/ but I am having a hard time tracking it down.~~

### CODE REVIEW

File: sys/parsoid.js [hunk 1]
```diff
@@ -587,6 +587,7 @@ class ParsoidService {
                 original
             };
             if (from === 'changes') {
+                req.body.changes = typeof req.body.changes === 'string' ? JSON.parse(req.body.changes) : req.body.changes;
```
Comment (Pchelolo) [inline sys/parsoid.js]: Why?

Comment (clarakosi) [reply sys/parsoid.js]: Done.

File: test/features/schema_tests.js [hunk 2]
```diff
@@ -26,28 +28,23 @@ parallel('Responses should conform to the provided JSON schema of the response',
...
     it('should expose valid OpenAPI spec', () => {
         return preq.get({ uri: `${server.config.baseURL()}/?spec` })
             .then((res) =>  {
-                return preq.post({uri: `http://online.swagger.io/validator/debug`,
-                    body: JSON.stringify(res.body)})
-                    .then((res) => {
-                        // if valid will return an empty object: {}
-                        assert.deepEqual({}, res.body, 'Spec must have no validation errors');
-                    })
+                assert.deepEqual({errors: []}, validator.validate(res.body), 'Spec must have no validation errors');
```
Comment (Pchelolo) [inline test/features/schema_tests.js]: Neat trick to show all errors nicely :)

**Comment by Pchelolo** [comment]
1. Needs a manual rebase
2. Needs to explicitly require a new version of hyperswitch once it's released with changes required for this
3. Let's not worry about travis for now, once @hknustwmf submits a PR for fixing travis (which is totally unrelated) we will rebase this.

**Comment by clarakosi** [comment]
Rebased but I still need to update service-checker to also look for the base URL at `servers[0].url`

**Comment by Pchelolo** [comment]
We are not going to deploy this very soon, so no rush

**Comment by d00rman** [comment]
@clarakosi I think we are finally getting ready to deploy this :) Could you (a) rebase on top of current master (a lot of things have changed!); and (b) bump hyperswitch' version to 0.12.2, please?

**Comment by d00rman** [comment]
> (b) bump hyperswitch' version to 0.12.2, please?

@clarakosi euh, this moved to v0.12.3 :P

**Comment by clarakosi** [comment]
@Pchelolo @d00rman Rebased + updated dependencies.

File: v1/announcements.yaml [hunk 3]
```diff
@@ -25,7 +24,8 @@ paths:
               gzip: false
             response:
               status: '{{get_announcements_from_mcs.status}}'
-              headers: '{{merge(get_announcements_from_mcs.headers, { "cache-control": options.announcement_cache_control})}}'
+              headers: '{{merge(get_announcements_from_mcs.headers, { "cache-control":
```
Comment (d00rman) [inline v1/announcements.yaml]: For clarity, let's not split the line

File: v1/mathoid.yaml [hunk 4]
```diff
@@ -122,38 +143,52 @@ paths:
...
       responses:
-        '200':
+        200:
           description: The rendered formula
-        '404':
+          content:
+            image/svg+xml:
+              schema:
+                type: string
+                format: binary
```
Comment (d00rman) [inline v1/mathoid.yaml]: SVG is not binary

File: v1/mathoid.yaml [hunk 5]
```diff
@@ -122,38 +143,52 @@ paths:
...
+            application/mathml+xml:
+              schema:
+                type: string
+                format: binary
```
Comment (d00rman) [inline v1/mathoid.yaml]: Idem

File: sys/parsoid.js [hunk 6]
```diff
@@ -77,7 +77,7 @@ function _dependenciesUpdate(hyper, req, newContent = true) {
     const rp = req.params;
     return mwUtil.getSiteInfo(hyper, req)
     .then((siteInfo) => {
-        const baseUri = siteInfo.baseUri.replace(/^https?:/, '');
+        const baseUri = siteInfo.baseUri.toString().replace(/^https?:/, '');
```
Comment (d00rman) [inline sys/parsoid.js]: I wonder if this should be fixed in `getSiteInfo` and stringified right away, given that it's the base URI. What do you think, @Pchelolo ?

Comment (Pchelolo) [reply sys/parsoid.js]: Ye, let's do it in the getSiteInfo. Agreed.

File: v1/announcements.yaml (hunk 3 above)
Comment (clarakosi) [reply v1/announcements.yaml]: Done.

File: v1/mathoid.yaml (hunk 4 above)
Comment (clarakosi) [reply v1/mathoid.yaml]: Done.

File: sys/parsoid.js (hunk 6 above)
Comment (clarakosi) [reply sys/parsoid.js]: Done.

File: v1/mathoid.yaml (hunk 5 above)
Comment (clarakosi) [reply v1/mathoid.yaml]: Done.

File: lib/mwUtil.js [hunk 7]
```diff
@@ -300,7 +300,10 @@ mwUtil.decodeBody = (contentResponse) => {
  */
 mwUtil.getSiteInfo = (hyper, req, domain) => hyper.get({
     uri: new URI([domain || req.params.domain, 'sys', 'action', 'siteinfo'])
-}).get('body');
+}).get('body').then((siteInfo) => {
```
Comment (Pchelolo) [inline lib/mwUtil.js]: Actually, let's go one step further and `toString` it in action.js?

**Comment by Pchelolo** [comment]
One small comment, otherwise it's great!

File: lib/mwUtil.js (hunk 7 above)
Comment (clarakosi) [reply lib/mwUtil.js]: Done.

**Comment by Pchelolo** [comment]
Great!

**Pull request merged**